File: julia/13_subsets.jl
# subsets(s,n): return the set of n-element subsets of s.

function subsets(s, n)
    # Base cases
    n == 0 && return [ [] ]           # one subset: the empty one
    n > length(s) && return []        # no way to pick n > |s|

    # Split off the first element
    head, tail = s[1], s[2:end]

    # 1) All subsets that include head: pick (n-1) from tail and prepend head
    with_head = []
    for sub in subsets(tail, n-1)
        push!(with_head, [head; sub])   # [head; sub] concatenates head in front
    end

    # 2) All subsets that exclude head: pick n from tail
    without = subsets(tail, n)

    # 3) Combine and return
    result = copy(with_head)
    for sub in without
        push!(result, sub)
    end
    return result
end

# Example
s = ['a','b','c','d']
println(subsets(s, 2))
# Any[Any['a', 'b'], Any['a', 'c'], Any['a', 'd'], Any['b', 'c'], Any['b', 'd'], Any['c', 'd']]

File: python/13_subsets.py
# subsets(s,n): return the set of n-element subsets of s.

# Leetcode 77. Combinations
# https://leetcode.com/problems/combinations/

def subsets(s, n):
    """
    Return a list of all n-element subsets of the list s.
    Each subset is a list; order of subsets does not matter.
    """
    # Base cases:
    if n == 0:
        return [[]]
    if n > len(s):
        return []

    head, tail = s[0], s[1:]

    # 1) Build all subsets that include `head`
    with_head = []
    # We need all (n-1)-element subsets of tail,
    # then for each one, prepend head.
    for comb in subsets(tail, n-1):
        # comb is one of the (n-1)-element subsets from tail
        new_subset = [head] + comb
        with_head.append(new_subset)
        # e.g. if head = 'a' and comb = ['c','d'], new_subset = ['a','c','d']

    # 2) Build all subsets that exclude `head`
    without_head = subsets(tail, n)

    # 3) Combine and return
    return with_head + without_head


# Example usage:
if __name__ == "__main__":
    s = ['a','b','c','d']
    print(subsets(s, 2))
    # [['a', 'b'], ['a', 'c'], ['a', 'd'], ['b', 'c'], ['b', 'd'], ['c', 'd']]

    print(subsets([1,2,3,4], 3))
    # [[1, 2, 3], [1, 2, 4], [1, 3, 4], [2, 3, 4]]
